src/coreclr/jit/redundantbranchopts.cpp

```diff
@@ -338,6 +501,67 @@ void Compiler::optRelopImpliesRelop(RelopImplicationInfo* rii)
                 }
             }
         }
+
+        // Given R(x, cns1) and R*(x, cns2) see if we can infer R* from R.
+        // We assume cns1 and cns2 are always on the RHS of the compare.
+        if ((treeApp.m_args[0] == domApp.m_args[0]) && vnStore->IsVNConstant(treeApp.m_args[1]) &&
+            vnStore->IsVNConstant(domApp.m_args[1]) && varTypeIsIntOrI(vnStore->TypeOfVN(treeApp.m_args[1])) &&
+            varTypeIsIntOrI(vnStore->TypeOfVN(domApp.m_args[1])))
+        {
+            // We currently don't handle VNF_relop_UN funcs here
+            if (ValueNumStore::VNFuncIsSignedComparison(domApp.m_func) &&
+                ValueNumStore::VNFuncIsSignedComparison(treeApp.m_func))
+            {
+                // Dominating "X relop CNS"
+                const genTreeOps domOper = static_cast<genTreeOps>(domApp.m_func);
+                const ssize_t    domCns  = vnStore->CoercedConstantValue<ssize_t>(domApp.m_args[1]);
+
+                // Dominated "X relop CNS"
+                const genTreeOps treeOper = static_cast<genTreeOps>(treeApp.m_func);
+                const ssize_t    treeCns  = vnStore->CoercedConstantValue<ssize_t>(treeApp.m_args[1]);
+
+                // Example:
+                //
+                // void Test(int x)
+                // {
+                //     if (x > 100)
+                //         if (x > 10)
+                //             Console.WriteLine("Taken!");
+                // }
+                //
+
+                // Corresponding BB layout:
+                //
+                // BB1:
+                //   if (x <= 100)
+                //       goto BB4
+                //
+                // BB2:
+                //   // x is known to be > 100 here
+                //   if (x <= 10) // never true
+                //       goto BB4
+                //
+                // BB3:
+                //   Console.WriteLine("Taken!");
+                //
+                // BB4:
+                //   return;
+
+                // Check whether the dominating compare being "false" implies the dominated compare is known
```

Comment:
The reversal/negation here is a bit tricky, but I think I follow it.

Seems like there's a complementary case where you don't negate the upper relop and then may instead have `canInferFromTrue` abilities?

Reply:
Yeah, basically there are two cases:

`canInferFromFalse`:

```cs
if (x > 100)
{
    if (x > 10)
    {
        ...
    }
}
```

`canInferFromTrue`:

```cs
if (x > 100)
{
    ...
}
if (x > 10)
{
    ...
}
```

I'll check separately the 2nd case (last I tried it added very little diffs)